### PR TITLE
Add a keepdim parameter for reduction functions over a single dimension.

### PR DESCRIPTION
**Pull request by gchanan**
By default, this parameter is False -- a backwards incompatible change, but
one that follows numpy semantics, e.g. numpy.sum (numpy names the parameter
"keepdims" since you can pass multiple dims to reduction functions).

The old behavior seems desired mainly for normalization type operations
where the tensor will immediately be expanded out again, e.g.:
probs.sum(1).expand_as(probs)
which no longer works as written because the dimension to expand is missing.
This can be fixed by simply passing True as "keepdim" argument
to the reduction operation, e.g:
probs.sum(1, keepdim=True).expand_as(probs)

### CODE REVIEW

**Comment by gchanan** [comment]
This addresses https://github.com/pytorch/pytorch/issues/289.

**Comment by apaszke** [comment]
I'm not 100% convinced about defaulting to what numpy does. It would break a lot of user code 😕

**Comment by soumith** [comment]
i want to get this into 0.2, along with broadcasting.
To have correct broadcasting without introducing user bugs, we have to introduce squeezing dims on reduction. And considering it's a major release (and the user can grep for relevant code easily / will get a hard-error), I think i prefer this.

**Comment by apaszke** [comment]
Why is it needed for correct broadcasting?

**Comment by soumith** [comment]
@apaszke see the comment here for why: https://github.com/pytorch/pytorch/issues/289#issuecomment-298705616

**Comment by apaszke** [comment]
I'm not entirely convinced by that comment either 😕

**Comment by gchanan** [comment]
Thanks for bringing up what the default should be @apaszke and @soumith.

I don't think the argument is really "To have correct broadcasting without introducing user bugs, we have to introduce squeezing dims on reduction", but rather "broadcasting is already going to break backwards compatibility, so let's pay the price once to get numpy-style semantics."

To wit:
- Broadcasting already breaks backwards compatibility -- i.e. a (1,4) x (4,1) tensor op becomes a (4,4) with broadcasting, compared to a (1,4) as we have now.
- As mentioned above, automatically squeezing the dimensions is also backwards incompatible, but there are cases (https://github.com/pytorch/pytorch/issues/289#issuecomment-298705616) where introducing it with broadcasting obviates the need for user-level changes compared to introducing broadcasting alone.  I don't have a good understanding for how prevalent this is, unfortunately.
- Getting close to numpy semantics (broadcasting) without going all the way (or most of the way, I'm sure we are missing some things) seems more confusing, i.e. if the difference between PyTorch and numpy semantics are totally different I can probably keep them straight, but if they are really close, it's actually more difficult.
- Removing the dimension by default is the consistent behavior: i.e. doing a reduction on a 1 dimension tensor without specify the dimension yields a scalar/0-dimensional tensor (torch.sum(x)) , but doing a reduction on a 1-dimensional tensor while specify the only dimensions yields a 1-dimensional tensor (torch.sum(x,0)).  Note that this requires https://github.com/pytorch/pytorch/issues/1433 to get totally consistent semantics.
- Making backwards incompatible changes are only going to get more difficult as usage of PyTorch increases and the code becomes more stable.  This, combined with us breaking backwards compatibility for broadcasting suggests this is the best time.

That being said, I understand the other side of the argument; as a user, dealing with backwards incompatibilities is a real pain.

One possibility is deferring the decision until broadcasting is ready.  It's not difficult to change the default at this point (now that I found the places that _need_ keepdims=True), so I could change the default for now, we can get this in, and we can re-evaluate with broadcasting.  That may give us more information on how prevalent the case mentioned in (https://github.com/pytorch/pytorch/issues/289#issuecomment-298705616) is.

Thoughts?

**Comment by gchanan** [comment]
The keepdim default is now True; the last two commits are split into:
- Changing the default
- Explicitly passing keepdim=False to tests that require it; i.e. can avoid reverting this if we change the default.

**Review by colesbury — APPROVED** [review]
LGTM.

I think in general it's makes the calling code more clear to use kwargs for `keepdim=True`, but if most of these call-sites are going away, it may not be worth making that change

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
this is now merged into master

**Comment by jekbradbury** [comment]
I'm a little confused -- I remembered the `keepdim` default as having changed to `False` and that's the behavior on current master but above it says "the keepdim default is now True". It's clear that `keepdim=False` is necessary for behavior like https://github.com/pytorch/pytorch/issues/289#issuecomment-298705616, but that wasn't possible in 0.1.12 so it's ok to require the kwarg. With `keepdim=False` as default the fairly common 0.1.12 patterns `var2 = var1.sum(dim); var2.expand_as(var1)` and `var2 = var1.sum(dim); var2.squeeze(dim)` both break, the latter without a Python traceback.

**Comment by gchanan** [comment]
The comment "the keepdim warning is now True" is not a typo; it was correct when it was written, i.e. it is explaining the context of the in progress commits, which first added a keepdim parameter (default True), then made the default False.

I edited the comment to explain that it is out of date to hopefully avoid future confusion.
